---
title: "RNAseq Quality Control"
author: "Inushi De Silva"
date: '2022-07-13'
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options("yaml.eval.expr" = TRUE)
```

**Repeat this process for *all cell lines*. Make sure to adjust the QC parameters as necessary.**

1.  ***First we will load the required libraries and set workingdirectories, datadirectories and plotdirectories.***

```{r Load Libraries and Set Directory Paths}
## Specify directories
script_path <- rstudioapi::getActiveDocumentContext()$path
script_dir <- dirname(script_path)
setwd(script_dir)
workingdirectory <- dirname(script_dir)
datadirectory <- paste0(workingdirectory, "/data")
resultsdirectory <- paste0(workingdirectory, "/results")
##dir.create(resultsdirectory)

## Load libraries
if(!require(dplyr)){
  install.packages("dplyr")
  library(dplyr)
}
if(!require(Seurat)){
  install.packages("Seurat")
  library(Seurat)
}
if(!require(ggplot2)){
  install.packages("ggplot2")
  library(ggplot2)
}
if(!require(sctransform)){
  install.packages("sctransform")
  library(sctransform)
}
if(!require(RColorBrewer)){
  install.packages("RColorBrewer")
  library(RColorBrewer)
}
if(!require(ggpointdensity)){
  install.packages("ggpointdensity")
  library(ggpointdensity)
}
```

2a. ***Next, read in the 10X feature matrices. This will be the path to a folder containing three .tgz files.***

```{r Load and Read 10x Data + Create Seurat Object}
##We set the path to the .tgz files
##Replace the name of the cell line each time
SANTB00111_TME_data <- paste0(datadirectory, "/Cellranger_outs/111_TME_feature_matrix/")
SANTB00111_CSF_data <- paste0(datadirectory, "/Cellranger_outs/111_CSF_feature_matrix/")

##We read in the 10x data
SANTB00111_TME <- Read10X(SANTB00111_TME_data)
SANTB00111_CSF <- Read10X(SANTB00111_CSF_data)

##Create separate seurat objects. 
Seurat_111_TME <- CreateSeuratObject(counts = SANTB00111_TME, project = "111_TME")
Seurat_111_CSF <- CreateSeuratObject(counts = SANTB00111_CSF, project = "111_CSF")

##Add metadata of PatientID and Conditions
Seurat_111_CSF$PatientID <- "SANTB00111"
Seurat_111_CSF$Condition <- "CSF"
Seurat_111_TME$PatientID <- "SANTB00111"
Seurat_111_TME$Condition <- "TME"
Seurat_111$Condition_PatientID <- paste(Seurat_111$PatientID, Seurat_111$Condition, sep = "_")

##If running on Mac, merging all Seurat objects may give the error - vector limit reached - to overcome this error make sure that you remove unnecessary Global Env and you may also go to Terminal, touch .Renv, open .Renv add R_MAX_VSIZE=100Gb and save. 
##We then merge CSF and TME together. This seurat object will be used for downstream processing. 
Seurat_111 <- merge(Seurat_111_CSF, y = Seurat_111_TME)

##Save Seurat object Pre-filter.
save(Seurat_111, file =  paste0(resultsdirectory, "/SANTB00111/RData/Prefilter_SANTB00111.RData/"))
```

2b. ***We visualise the spread of the data, number of cells, number of genes and number of reads per cell here. We calculate percentages of mitochondrial reads for each cell and also calculate the number of housekeeping genes expressed by each cell. This will help us filter out any dead cells or empty droplets. Run this and replace name of Seurat Object each time you want to visualise a different cell iine***

```{r Data Visualisation and QC}
########## PRE-PROCESSING DATA QUALITY CONTROL & FILTERING #################
##NOTE: In each seurat object, the nCountRNA = Number of UMIs per cell and nFeature_RNA = number of genes detected per cell
## Letâ€™s quickly do some plotting to look at the number of reads per cell and detected/expressed genes per cell (often called complexity)

## Plot a frequency distribution (histogram) of the number of reads
hist <- ggplot(Seurat_111@meta.data, aes(Seurat_111@meta.data$nCount_RNA)) + theme_light() +
  geom_histogram(bins = 50, color = "black", fill = "grey") +
  theme(legend.position = "none", axis.text.x = element_text(size = 16, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 16), axis.title.x = element_text(size = 16), axis.title.y = element_text(size = 16), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of Reads") + ylab("Number of Cells") +
  scale_x_continuous(breaks = seq(0, 75000, by = 15000)) +
  scale_y_continuous(breaks = seq(0, 1000, by = 250)) +
  coord_cartesian(xlim = c(0, 75000), ylim = c(0, 1000)) + geom_vline(xintercept = 55000, linetype = "dashed", col = "red", size = 1)
hist
ggsave(hist, file = paste(resultsdirectory, "SANTB00111/plots/Seurat_111_reads_per_cell_beforefiltering.pdf", sep = "/"), width = 10, height = 6)


## Plot a frequency distribution (histogram) of the number of detected genes
hist2 <- ggplot(Seurat_111@meta.data, aes(Seurat_111@meta.data$nFeature_RNA)) + theme_light() + 
  geom_histogram(bins = 50, color = "black", fill = "grey") +
  theme(legend.position = "none", axis.text.x = element_text(size = 16, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 16), axis.title.x = element_text(size = 16), axis.title.y = element_text(size = 16), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of Genes Expressed") + ylab("Number of Cells") +
  scale_x_continuous(breaks = seq(0, 9000, by = 1000)) +
  scale_y_continuous(breaks = seq(0, 750, by = 250)) +
  coord_cartesian(xlim = c(0, 9000), ylim = c(0, 750)) + geom_vline(xintercept = 2000, linetype = "dashed", col = "red", size = 1)
hist2
ggsave(hist2, file = paste(resultsdirectory, "SANTB00111/plots/Seurat_111_genes_per_cell_beforefiltering.pdf", sep = "/"), width = 10, height = 6)
```

3.  ***Following Visualisation we add new metadata for mitochondrial reads and housekeeping genes***

```{r Add Metadata - Mitochondrial reads, Housekeeping Genes}
## Filter and Create New Seurat Object  #####
SANTB00111 <- CreateSeuratObject(counts = Seurat_111@assays$RNA@counts, meta.data = Seurat_111@meta.data, min.cells = 3, min.features = 200) ## include only genes that are are expressed in 3 or more cells and cells with complexity of 200 genes or more
SANTB00111@assays[["RNA"]]@counts[1:10,1:10] # look at the first 10 genes (rows) and first 10 cells (columns)
dim(SANTB00111) ## 21877 genes and 6896 cells
dim(Seurat_111) ##36601 genes and 6933 cells

## Compute the proportion of reads that are of mitochondrial origin for every cell (percent.mito)
SANTB00111 <- PercentageFeatureSet(object = SANTB00111, pattern = "^MT-", col.name = "percent.mt") ## store mitochondrial perentage in object meta data
head(SANTB00111[["percent.mt"]])
head(SANTB00111@meta.data, 6) ## show QC metrics for the first 6 cells

## Compute the number of house keeping genes expressed in each cell -- HKGs are abundant and are usually steadliy expressed in cells, thus less sensitive to the high dropout
## housekeepers.txt: list of 98 housekeeping genes compiled in Tirosh et al., 2016, to be used in data preprocessing, to remove sources of unwanted variation -- downloaded from 
hkgenes <- read.table(paste0(datadirectory, "/reference_sheets/housekeepers.txt"))
## load the the list of house keeping genes
hkgenes <- as.vector(hkgenes$V1)
hkgenes.found <- which(toupper(rownames(SANTB00111@assays[["RNA"]]@counts)) %in% hkgenes) ## remove hkgenes that were not found
length(hkgenes.found) ## 92/98 HKGs found
n.expressed.hkgenes <- Matrix::colSums(SANTB00111@assays[["RNA"]]@counts[hkgenes.found, ] > 0) # sum the number of detected house keeping genes for each cell, then add this to the meta data
SANTB00111 <- AddMetaData(object = SANTB00111, metadata = n.expressed.hkgenes, col.name = "n.exp.hkgenes")
```

4.  ***Visualisation of Mitochondrial Reads and Housekeeping Genes - For each cell line, run this by replacing name of Seurat Object each time***

```{r Visualisation of Mitochondrial Reads and Housekeeping Genes}
## Visualize these QC metrics
VlnPlot(object = SANTB00111, features = c("nFeature_RNA", "nCount_RNA", "percent.mt","n.exp.hkgenes"), ncol = 4,  pt.size = 0)
ggsave(file = paste(resultsdirectory, "SANTB00111/plots/SANTB00111_seurat_QC_metrics_beforefiltering.pdf", sep = "/"), width = 10, height = 6)

##First we visualise the nCount_RNA against nFeature_RNA prior to filtering the SANTB00111 data. 
##Run the same line following filtering in Step 8. 
scatterPlot <- ggplot(SANTB00111@meta.data, aes(x = nCount_RNA, y = nFeature_RNA)) +
  theme_light() +
  geom_point(size = 2) +
  geom_pointdensity(adjust = 0.5) +
  scale_colour_gradientn(colours = colorRampPalette(rev(brewer.pal(11,'Spectral')))(8)) + ## colors in the scale
  theme(legend.position = "right", axis.text.x = element_text(size = 22, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 22), axis.title.x = element_text(size = 22), axis.title.y = element_text(size = 22), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of UMI Counts") + ylab("Number of genes detected") +
  scale_x_continuous(breaks = seq(0, 75000, by = 15000)) +
  scale_y_continuous(breaks = seq(0, 9000, by = 1500)) +
  coord_cartesian(xlim = c(0, 75000), ylim = c(0, 9000)) +
  geom_hline(yintercept = 2000, linetype = "dashed", col = "red", size = 1) + 
  geom_vline(xintercept = 55000, linetype = "dashed", col = "red", size = 1)
scatterPlot
ggsave(scatterPlot, filename = paste(resultsdirectory, "SANTB00111/plots/PreQC_SANTB00111_UMIcounts_versus_genesdetected_per_cell_legend.pdf", sep = "/"), height = 6, width = 6, dpi = 300, useDingbats = FALSE)

##Next we visualise the number of genes dected against the percentage of mitochondrial reads prior to filtering the SANTB00111 data.
scatterPlot <- ggplot(SANTB00111@meta.data, aes(x = nFeature_RNA, y = percent.mt)) +
  theme_light() +
  geom_point(size = 2) +
  geom_pointdensity(adjust = 0.5) +
  scale_colour_gradientn(colours = colorRampPalette(rev(brewer.pal(11,'Spectral')))(8)) + ## colors in the scale
  theme(legend.position = "none", axis.text.x = element_text(size = 22, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 22), axis.title.x = element_text(size = 22), axis.title.y = element_text(size = 22), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of Genes Expressed") + ylab("% of Mitochondrial Reads") +
  scale_x_continuous(breaks = seq(0, 9000, by = 1500)) +
  scale_y_continuous(breaks = seq(0, 100, by = 10)) +
  coord_cartesian(xlim = c(0, 9000), ylim = c(0, 100)) +
  geom_hline(yintercept = 20, linetype = "dashed", col = "red", size = 1) + 
  geom_vline(xintercept = 2000, linetype = "dashed", col = "red", size = 1)
scatterPlot
ggsave(scatterPlot, filename = paste(resultsdirectory, "SANTB00111/plots/PreQC_SANTB00111_genesdetected_versus_percentmtreads_per_cell_nolegend.pdf", sep = "/"), height = 6, width = 6, dpi = 300, useDingbats = FALSE)

##Do the same with nCountRNA.
scatterPlot <- ggplot(SANTB00111@meta.data, aes(x = nCount_RNA, y = percent.mt)) +
  theme_light() +
  geom_point(size = 2) +
  geom_pointdensity(adjust = 0.5) +
  scale_colour_gradientn(colours = colorRampPalette(rev(brewer.pal(11,'Spectral')))(8)) + ## colors in the scale
  theme(legend.position = "none", axis.text.x = element_text(size = 22, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 22), axis.title.x = element_text(size = 22), axis.title.y = element_text(size = 22), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of UMI Counts") + ylab("% of Mitochondrial Reads") +
  scale_x_continuous(breaks = seq(0, 75000, by = 15000)) +
  scale_y_continuous(breaks = seq(0, 100, by = 10)) +
  coord_cartesian(xlim = c(0, 75000), ylim = c(0, 100)) +
  geom_hline(yintercept = 20, linetype = "dashed", col = "red", size = 1) + 
  geom_vline(xintercept = 55000, linetype = "dashed", col = "red", size = 1)
scatterPlot
ggsave(scatterPlot, filename = paste(resultsdirectory, "SANTB00111/plots/PreQC_SANTB00111_umicounts_versus_percentmtreads_per_cell_nolegend.pdf", sep = "/"), height = 6, width = 6, dpi = 300, useDingbats = FALSE)

## Plot of # of HKG genes expressed VS # of genes detected. 
scatterPlot <- ggplot(SANTB00111@meta.data, aes(x = nFeature_RNA, y = n.exp.hkgenes)) +
  theme_light() +
  geom_point(size = 2) +
  geom_pointdensity(adjust = 0.5) +
  scale_colour_gradientn(colours = colorRampPalette(rev(brewer.pal(11,'Spectral')))(8)) +
  theme(legend.position = "none", axis.text.x = element_text(size = 22, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 22), axis.title.x = element_text(size = 22), axis.title.y = element_text(size = 22), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of genes detected") + ylab("HK genes expressed") +
  scale_x_continuous(breaks = seq(0, 9000, by = 1500)) +
  scale_y_continuous(breaks = seq(0, 100, by = 10)) +
  coord_cartesian(xlim = c(0, 9000), ylim = c(0, 100)) +
  geom_hline(yintercept = 70, linetype = "dashed", col = "red", size = 1) + 
  geom_vline(xintercept = 2000, linetype = "dashed", col = "red", size = 1)
scatterPlot
ggsave(scatterPlot, filename = paste(resultsdirectory, "SANTB00111/plots/PreQC_SANTB00111_genesdetected_versus_nexpHKgenes_per_cell_nolegend.pdf", sep = "/"), height = 6, width = 6, dpi = 300, useDingbats = FALSE)

##Do the same with nCountRNA.
scatterPlot <- ggplot(SANTB00111@meta.data, aes(x = nCount_RNA, y = n.exp.hkgenes)) +
  theme_light() +
  geom_point(size = 2) +
  geom_pointdensity(adjust = 0.5) +
  scale_colour_gradientn(colours = colorRampPalette(rev(brewer.pal(11,'Spectral')))(8)) +
  theme(legend.position = "none", axis.text.x = element_text(size = 22, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 22), axis.title.x = element_text(size = 22), axis.title.y = element_text(size = 22), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
  xlab("Number of UMI counts") + ylab("HK genes expressed") +
  scale_x_continuous(breaks = seq(0, 75000, by = 15000)) +
  scale_y_continuous(breaks = seq(0, 100, by = 10)) +
  coord_cartesian(xlim = c(0, 75000), ylim = c(0, 100)) +
  geom_hline(yintercept = 70, linetype = "dashed", col = "red", size = 1) + 
  geom_vline(xintercept = 55000, linetype = "dashed", col = "red", size = 1)
scatterPlot
ggsave(scatterPlot, filename = paste(resultsdirectory, "SANTB00111/plots/PreQC_SANTB00111_umicounts_versus_nexpHKgenes_per_cell_nolegend.pdf", sep = "/"), height = 6, width = 6, dpi = 300, useDingbats = FALSE)
```

5.  ***Now we filter out any dead cells and cells with low numbers of HK genes***

```{r QC of Bardy Seurat Object}
## Filter the cells based on the quality control metrics: filtering based on (1) # of detected genes, (2) percent mitochondrial reads, (3) no. of housekeeping genes expressed
SANTB00111_QC <- subset(x = SANTB00111, subset = nFeature_RNA >= 2000 & percent.mt < 20 & n.exp.hkgenes >= 70 & nCount_RNA < 55000)
## We will remove cells with < 2000 genes detected, with 20% or more mitochondrial reads, and with less than 70 hKG expressed
dim(SANTB00111_QC) ## 21877 genes and 6381 cells AFTER filtering, After QC
dim(SANTB00111) ## 21877 genes and 6896 cells BEFORE filtering, PreQC

## Make histogram of number of cells against # of genes detected
##Run the same line following filtering in Step 8. 
histoPlot <- ggplot(SANTB00111_QC@meta.data, aes(nFeature_RNA)) +
  theme_light() +
  geom_histogram(bins = 50, fill = "grey", col = "black") +
  xlab("Number of Genes Expressed") + ylab("Number of cells") +
  theme(legend.position = "none", axis.text.x = element_text(size = 16, angle = 90, hjust = 1, vjust = 0.5), axis.text.y = element_text(size = 16), axis.title.x = element_text(size = 16), axis.title.y = element_text(size = 16), panel.border = element_rect(fill=NA, colour = "black", size=1.5)) +
scale_x_continuous(breaks = seq(0, 9000, by = 1000)) +
 scale_y_continuous(breaks = seq(0, 1000, by = 250)) +
 coord_cartesian(xlim = c(0, 9000), ylim = c(0, 1000)) +
 geom_vline(xintercept = 2000, linetype = "dashed", col = "red", size = 1)
histoPlot
ggsave(histoPlot, filename = paste(resultsdirectory, "SANTB00111/plots/SANTB00111_histogram_genes_per_cell_preQC.pdf", sep = "/"), height = 6, width = 10, dpi = 300, useDingbats = FALSE)

##We subset the QC'd seurat object by CSF and TME and save. We will use these objects to create a merged dataset. 
SANTB00111_CSF <- subset(x = SANTB00111_QC, subset = Condition == "CSF")
SANTB00111_TME <- subset(x = SANTB00111_QC, subset = Condition == "TME")
save(SANTB00111_CSF, file = paste(datadirectory, "/RData/Individual Cell Line RData/SANTB00111_CSF.Rdata"))
save(SANTB00111_TME, file = paste(datadirectory,"/RData/Individual Cell Line RData/SANTB00111_TME.Rdata"))

##Save Seurat objects QC and following filtering.
save(SANTB00111, file = paste(resultsdirectory,"/SANTB00111/RData/SANTB00111_PreQC.Rdata"))
save(SANTB00111_QC, file = paste(resultsdirectory,"/SANTB00111/RData/SANTB00111_QC.Rdata"))
```

6.  ***Data visualisation of cells Pre-QC***

```{SUPPLEMENTARY FIGURE 10B and C}
## We first load all the pre-filtered Seurat Objects for each cell line.
## We then merge them into a Seurat Object - Bardy_10X. 
Bardy_10x <- merge(Seurat_111, y = c(Seurat_134, Seurat_159, Seurat_448, Seurat_468, Seurat_469, Seurat_497, Seurat_BAH1, Seurat_BAH1B, Seurat_HW1, Seurat_MN1))

##We extract the metadata and the save the seurat object to generate violin plots in PRISM.
write.csv(Bardy_10x@meta.data, file = paste(datadirectory,"/RData/metadata sheets/Prefilter_GBM_MetaData.csv"))
save(Bardy_10x, file = paste(datadirectory,"/RData/Prefilter_GBM_All.Rdata"))
```
